src/msg_handler.c

```diff
@@ -0,0 +1,388 @@
+                int_button_pressed = false;
+                ext_button_pressed = false;
+                return strlen(response);
+            }
+
+            strncat(response, "NACK:Invalid button command", max_len);
+            return strlen(response);
+        }
+        strncat(response, "NACK:Invalid button name", max_len);
+        return strlen(response);
+    } else if (strcmp(next_arg, "NOTE") == 0) {
+        next_arg = strtok(NULL, ":");
+        if (next_arg == NULL) {
+            strncat(response, "NACK:Missing note frequency", max_len);
+            return strlen(response);
+        }
+
+        int note_freq;
+
+        if (isdigit((int)next_arg[0])) {
+            note_freq = atoi(next_arg);
+            // bounds check
+            if (note_freq < 0) {
+                strncat(response, "NACK:Invalid note frequency", max_len);
+                return strlen(response);
+            }
+        } else {
+            strncat(response, "NACK:Invalid note frequency", max_len);
+            return strlen(response);
+        }
+
+        next_arg = strtok(NULL, ":");
+        if (next_arg == NULL) {
+            strncat(response, "NACK:Missing note duration", max_len);
+            return strlen(response);
+        }
+
+        int note_dur;
+
+        if (isdigit((int)next_arg[0])) {
+            note_dur = atoi(next_arg);
+            // bounds check
+            if (note_dur < 0) {
+                strncat(response, "NACK:Invalid note duration", max_len);
+                return strlen(response);
+            }
+        } else {
+            strncat(response, "NACK:Invalid note duration", max_len);
+            return strlen(response);
+        }
+
+        // Generate note
+        buzzer_note(note_freq, note_dur);
+
+        strncat(response, "ACK", max_len);
+        return strlen(response);
+    } else if (strcmp(next_arg, "*IDN?") == 0) {
+        strncat(response, "Student Robotics:", max_len);
+        strncat(response, BOARD_NAME_SHORT, max_len - strlen(response));
+        strncat(response, ":", max_len - strlen(response));
+        strncat(response, (const char *)SERIALNUM_BOOTLOADER_LOC, max_len - strlen(response));
+        strncat(response, ":", max_len - strlen(response));
+        strncat(response, FW_VER, max_len - strlen(response));
+        return strlen(response);
+    } else if (strcmp(next_arg, "*STATUS?") == 0) {
+        for (output_t out=OUT_H0; out <= OUT_5V; out++) {
+            strncat(response, (output_inhibited[out])?"1,":"0,", max_len - strlen(response));
+        }
+        uint8_t final_char = strlen(response) - 1;
+        if (response[final_char] == ',') {  // clear final comma
+            response[final_char] = '\0';
+        }
+        strncat(response, ":", max_len - strlen(response));
+        strncat(response, itoa(board_temp, temp_str), max_len);
+        strncat(response, ":", max_len - strlen(response));
+        strncat(response, (fan_running()?"1":"0"), max_len);
```

Comment:
```suggestion
        strncat(response, itoa(board_temp, temp_str), max_len - strlen(response));
        strncat(response, ":", max_len - strlen(response));
        strncat(response, (fan_running()?"1":"0"), max_len - strlen(response));
```

I'd strongly recommend wrapping these calls in a function like
```c
static void append_str(char *dest, int dest_max_len, const char *src) {
  strncat(dest, src, dest_max_len - strlen(dest));
}
```
so that it's much harder to make this mistake.

Reply:
I've added the `append_str` function and `get_next_arg` function to abstract some of the string handling.